review09/README.md

```diff
@@ -0,0 +1,197 @@
+- `@Retry` -- It specifies that a method should be executed again if it fails. You can specify the maximum number of
+  retries and the delay between retries. It's useful for handling transient failures.
+- `@CircuitBreaker` -- It specifies that a method should be executed in a circuit breaker. If the method fails, the
+  circuit is opened, and the method is not executed for a specified time. After that time, the circuit is half-opened,
+  and the method is executed again. If it fails, the circuit is opened again. If it succeeds, the circuit is closed
+  again. It's useful for not overloading a failing service.
+- `@Fallback` -- It specifies a fallback method that should be executed if the original method fails. It's useful for
+  providing a default value (e.g. default recommendations) or a default behavior when a method fails.
+
+#### Example
+
+```java
+public class HelloResource {
+
+    @GET
+    @Path("/hello")
+    @Timeout(1000)
+    public String hello() {
+        // calling an external service for e.g. name
+    }
+
+    @GET
+    @Path("/hello")
+    @Retry(maxRetries = 3, delay = 1000)
+    public String hello() {
+        // calling an external service for e.g. name
+    }
+
+    @GET
+    @Path("/hello")
+    @CircuitBreaker(
+            requestVolumeThreshold = 4, // number of requests before the circuit is opened 
+            failureRatio = 0.75,  // the failure ratio at which the circuit should open
+            delay = 1000,  // the time the circuit should be opened before it's half-opened
+            successThreshold = 2  // the number of successful requests before the circuit is closed again
+    )
+    public String hello() {
+        // calling an external service for e.g. name
+    }
+
+    @GET
+    @Path("/hello")
+    @Fallback(fallbackMethod = "fallbackHello")
+    public String hello() {
+        // calling an external service for e.g. name
+    }
+
+    public String fallbackHello() {
+        return "Hello, fallback!";
+    }
+}
+```
+
+## Tracing
+
+In modern cloud-based applications, it is important to have visibility into the system. Tracing is a technique to follow the flow of the code and understand how the different services interact with each other. It is useful for debugging, performance analysis, and understanding the system.
+
+### OpenTelemetry
+
+OpenTelemetry is a set of APIs, libraries, agents, and instrumentation to provide observability for cloud-native software. It provides a set of tools to generate, collect, and export telemetry data (traces) for analysis.
+
+### Jaeger
+
+Jaeger is an open-source, end-to-end distributed tracing system. It is used to monitor and troubleshoot microservices-based distributed systems. It is used to monitor and troubleshoot distributed systems.
+
+It collects data from the services using OpenTelemetry over gRPC, stores them, and visualizes the traces in a user-friendly way.
```

Comment:
OTLP is also possible over HTTP 1.1, not only gRPC - https://opentelemetry.io/docs/specs/otlp/#otlphttp

Reply:
also https://www.jaegertracing.io/docs/1.21/opentelemetry/ 

Enables Jaeger receiver (by default) with collector endpoints - gRPC, TChannel, HTTP.